pilot/pkg/proxy/envoy/config.go

```diff
@@ -134,23 +134,33 @@ func BuildConfig(config meshconfig.ProxyConfig, pilotSAN []string) *Config {
 
 // buildListeners produces a list of listeners and referenced clusters for all proxies
 func buildListeners(env model.Environment, node model.Node) (Listeners, error) {
+	var listeners Listeners
+
 	switch node.Type {
-	case model.Sidecar, model.Router:
+	case model.Sidecar:
 		instances, err := env.HostInstances(map[string]*model.Node{node.IPAddress: &node})
 		if err != nil {
 			return nil, err
 		}
+
 		services, err := env.Services()
 		if err != nil {
 			return nil, err
 		}
-		listeners, _ := buildSidecarListenersClusters(env.Mesh, instances,
-			services, env.ManagementPorts(node.IPAddress), node, env.IstioConfigStore)
-		return listeners, nil
+
+		listeners, _ = buildSidecarListenersClusters(env.Mesh, instances, services, env.ManagementPorts(node.IPAddress), node, env.IstioConfigStore)
+	case model.Router:
+		instances, err := env.HostInstances(map[string]*model.Node{node.IPAddress: &node})
+		if err != nil {
+			return nil, err
+		}
+
+		listeners = buildGatewayListeners(env.Mesh, instances, env.IstioConfigStore, node)
 	case model.Ingress:
-		return buildIngressListeners(env.Mesh, nil, env.ServiceDiscovery, env.IstioConfigStore, node), nil
+		listeners = buildIngressListeners(env.Mesh, nil, env.ServiceDiscovery, env.IstioConfigStore, node)
```

Comment:
FYI: The plan is to merge model.Ingress with model.Gateway, such that ingress is a special purpose gateway whose configuration flows in via kubernetes. Will leave it to you guys to take that in one PR or multiple

Reply:
Can you elaborate on this a bit - does that mean we should always be merging the listeners of `model.Ingress` & `model.Gateway` if we have them both?

Reply:
Ignore for the moment please. @ZackButcher may be able to shed more light.